MDGPS

wmontgomery4 opened the pull request:
Sorry for the delay, here's a preliminary pull request for the mdgps code. It isn't in a state to be merged just yet, but I wanted to let you take a look at the main outline. The standing issues right now are:
1) Some of the changes are pretty hacky (particularly to get on-policy sampling working), and should be made more sane before merging. This also includes adding a separate traj_opt file for MDGPS, and the way that I use a 'pol_info' object as a 'traj_distr' object sometimes.
2) I didn't change any of the tf policy_opt code, but did make changes to the caffe policy_opt code, so things probably aren't compatible between the two right now
3) On-policy sampling isn't working that well with just 4 conditions on the peg insertion task (we used 9 in the paper), so I need to play around with that and see if it's just a hyperparameter issue.
4) There are also some minor tweaks to the gui and some fixed TODOs from non-mdgps stuff, which I should probably just open a new branch for

I'm working on the issues above, but feel free to take a look over things in the meantime.

wmontgomery4 committed:
set up mdgps code/experiment

wmontgomery4 committed:
fixed kl_step_schedule

wmontgomery4 committed:
Merge branch 'master' into mdgps

wmontgomery4 committed:
made mdgps work with peg fix

wmontgomery4 committed:
cleaned up algorithm_md to remove dual variable stuff

wmontgomery4 committed:
sampling from nn policy switch

wmontgomery4 committed:
cleaned some stuff up, added vanilla step adjustment

wmontgomery4 committed:
added policy costs to gui output

wmontgomery4 committed:
fixed step_mult bug, switched to prev=nn, cur=lqr

wmontgomery4 committed:
debug 2 condition expts and logs

wmontgomery4 committed:
added random seed expts for lqr, badmm, mdgps

wmontgomery4 committed:
turned off NN sampling on mjc_mdgps_example

wmontgomery4 committed:
added avg_pol_cost to GUI, added big peg expts

wmontgomery4 committed:
fixed peg_big examples, added square peg example

wmontgomery4 committed:
added peg3D dir

wmontgomery4 committed:
cleaned some stuff up, added plotting code

wmontgomery4 committed:
added new step size rule flag

wmontgomery4 committed:
added peg3D_hard/rollingball tasks

wmontgomery4 committed:
added 'weighted_kl' flag, removed some old expts

wmontgomery4 committed:
added seeded runs for peg3D/peg3D_hard/rollingball

wmontgomery4 committed:
made gmm converge faster

wmontgomery4 committed:
rollingball not working

wmontgomery4 committed:
cleanup, got rid of new step size expts

wmontgomery4 committed:
added some new debug expts/flags

wmontgomery4 committed:
made debug hard grid smaller, added more expts

wmontgomery4 committed:
fixed rollingball

wmontgomery4 committed:
made debug grid smaller, added new step expts

wmontgomery4 committed:
moved some stuff around, changed run_seeds.sh, ready for reacher3 suite

wmontgomery4 committed:
changed reacher3 hp, some cleanup

wmontgomery4 committed:
lots of stuff

wmontgomery4 committed:
so many things... trying to fix reacher3 now

wmontgomery4 committed:
more fixes?

wmontgomery4 committed:
again, so much stuff, gotta be safe

wmontgomery4 committed:
expts are basically done, try another obstacle task

wmontgomery4 committed:
in submission state, needs quite a bit of clean up

wmontgomery4 committed:
changes from chelseas comments

wmontgomery4 committed:
initial cleanup of nips expts

wmontgomery4 committed:
Merge branch 'mdgps'

wmontgomery4 committed:
more cleanup

wmontgomery4 committed:
change step_rule to classic/global from old/new

wmontgomery4 committed:
setup mjc_mdgps_example variations

wmontgomery4 committed:
removed lg_covar stuff

wmontgomery4 committed:
removed weighted_kl stuff

wmontgomery4 committed:
get rid of return_sum option in traj_distr_kl

cbfinn commented:
Thanks Harley! I'll take a look at it now, and go ahead and repost/comment when you make changes that are ready to be looked at.

cbfinn commented on mjc_models/obstacle_course.xml:
```diff
@@ -0,0 +1,41 @@
+<mujoco model="linearparticle">
```
Since there are now a fair number of xml files, it would be nice to have comments in this file and/or a screen shot of the environment for each xml file.

cbfinn commented on python/gps/algorithm/algorithm_mdgps.py:
```diff
@@ -0,0 +1,341 @@
+            self._eval_cost(m)
+            lmax = self.cur[m].cs.max(axis=0)
+            lmin = self.cur[m].cs.min(axis=0)
+            self.cur[m].qmax = np.cumsum( (lmax - lmin)[::-1] )[::-1]
```
what is this code doing? [maybe add a explanatory comment?]

wmontgomery4 replied:
This was for an idea Sergey had where we compute the cumulative Qmax (like in the bound from the paper), in order to try and inform the step size updates. We never got it working, and I forgot to remove it.

cbfinn commented on python/gps/algorithm/algorithm_mdgps.py:
```diff
@@ -0,0 +1,341 @@
+                self._stepadjust(m)
+
+    def _update_policy(self, itr, inner_itr):
+        """ Compute the new policy. """
```
It would be nice if we could reuse more code between this file and algorithm_badmm. (Not a super high priority though)

cbfinn commented on python/gps/algorithm/traj_opt/traj_opt_lqr_python.py:
```diff
@@ -244,7 +253,7 @@ def backward(self, prev_traj_distr, traj_info, eta, algorithm, m):
                     break
 
-                # Store conditional covariance, inverse, and Cholesky.
+                # Store conditional covariance, inverse, and Cholesky
```
nitpic: periods are nice (and are consistent with pylint standards).

cbfinn commented on python/gps/algorithm/traj_opt/traj_opt_utils.py:
```diff
@@ -85,7 +85,7 @@ def traj_distr_kl(new_mu, new_sigma, new_traj_distr, prev_traj_distr):
 
     # Add up divergences across time to get total divergence.
-    return np.sum(kl_div)
+    return kl_div
```
It seems like this change would break something in traj_opt_lqr.py... Was this a bug in the code?

wmontgomery4 replied:
No, this was a mistake I made when setting up the pull request, along with several of the other comments. I had some code supporting weighted KL constraints but took it out since it wasn't helping and cluttered things. I'm making a fix for that now.

wmontgomery4 committed:
fixes from chelseas comments

wmontgomery4 committed:
fixed other comment period

wmontgomery4 committed:
added comment to obstacle course, removed reacher3/rollingball

wmontgomery4 committed:
Merge branch 'master' of github.com:cbfinn/gps

wmontgomery4 committed:
Merge branch 'master' of github.com:cbfinn/gps

wmontgomery4 committed:
minor tweaks to mjc_mdgps_example

wmontgomery4 committed:
wip

wmontgomery4 committed:
unroll 2 iterations

wmontgomery4 committed:
unroll 1 iteration, working for off_classic, almost for on_global

wmontgomery4 committed:
wip

wmontgomery4 committed:
fixed inner_itr stuff, working with one inner iteration

wmontgomery4 committed:
hyperparam tweaking, back to big grid, nearly working for all 4

wmontgomery4 committed:
remove unused square peg xml

wmontgomery4 committed:
runs with tf, but works worse and is way slower

wmontgomery4 committed:
removed some junk

wmontgomery4 committed:
agent_use_nn_policy -> sample_on_policy, minor hp tweaks

wmontgomery4 committed:
move mdgps traj_opt stuff to one file

wmontgomery4 commented:
Okay, this is in a state to be looked at again. Couple notes:
- Works with tf as well, but it's slower and doesn't seem to work quite as well (although this could easily be a fluke, or because the hyperparams were tweaked with caffe). Tensorflow also takes up a lot of memory, so I can't run all 4 variations simultaneously, so I'm sticking with caffe for now, but it works with tensorflow and so that's not an issue.
- I had to remove the default TrajOptLqrPython from the algorithm config.py to avoid circular imports (I need to import both AlgorithmBADMM and AlgorithmMDGPS in TrajOpt for some type checks since the traj opt varies slightly between them)
- Still haven't refactored things with BADMM, but I agree that would be a good next step.

cbfinn commented on python/gps/algorithm/algorithm_mdgps.py:
```diff
@@ -0,0 +1,334 @@
+
+    def iteration(self, sample_lists):
+        """
+        Run iteration of BADMM-based guided policy search.
```
change BADMM to mirror descent

cbfinn commented on python/gps/algorithm/algorithm_mdgps.py:
```diff
@@ -0,0 +1,334 @@
+            self.new_traj_distr = [
+                self.cur[cond].traj_distr for cond in range(self.M)
+            ]
+            self._update_policy(self.iteration_count, 1) # HACK: need to set inner_itr=1
```
Why is this a hack? I guess it would be really easy to change the policy_opt_caffe code to instead take in a normalize=True/False flag, or check if the policy scale and bias already exists. Could you make this change? [It will also involve a small change to algorithm_badmm]

wmontgomery4 replied:
It's a bit of a hack because it's not really inner_itr 1, you just have to say that so that the mean and variance will be set correctly. Agreed that adding a normalize flag is the better option, and I'll make that change. In the future we might be better off just using batch norm, especially since the conditions will be changing randomly, but let's put that off for now.

wmontgomery4 replied:
Ended up just checking whether or not the scale/bias exists already. Seemed like an easier option for now, which could be replaced with batch norm later.

cbfinn replied:
Sounds good. I'd rather use a variant of batch norm which does not rely on the statistics of the batch, since we need to evaluate single samples effectively. There are a couple of options in this direction.

cbfinn commented on python/gps/algorithm/algorithm_mdgps.py:
```diff
@@ -0,0 +1,334 @@
+        # TODO: store init_kl after updating the trajectory
+        self._update_step_size()  # KL Divergence step size (also fits policy).
+        self._update_trajectories()
+        self._update_policy(self.iteration_count, 1) # HACK: need to set inner_itr=1
```
why do you set inner_itr to 1 here? (why recompute the mean and variance?)

wmontgomery4 replied:
Think this was just a c&p error, no need to recompute the mean/variance as you say.

cbfinn commented on python/gps/algorithm/algorithm_mdgps.py:
```diff
@@ -0,0 +1,334 @@
+
+    def _update_policy_samples(self):
+        """ Update the list of samples to use with the policy. """
+        #TODO: Handle synthetic samples.
```
Do you plan to add synthetic samples? If not, let's remove this TODO.

wmontgomery4 replied:
No, this was copied and pasted from the BADMM code, should I remove the TODO there as well?

cbfinn replied:
Let's keep it in BADMM for now. I'm not sure if we'll add it for BADMM, probably not, but nice to be aware of.

cbfinn commented on python/gps/algorithm/config.py:
```diff
@@ -18,13 +15,13 @@
     'init_traj_distr': None,  # A list of initial LinearGaussianPolicy
                               # objects for each condition.
     # Trajectory optimization.
-    'traj_opt': TrajOptLQRPython({}),
```
I think this is okay. In practice, we'll always specify this in the hyperparams anyway.

cbfinn commented on python/gps/algorithm/config.py:
```diff
@@ -44,3 +41,12 @@
+
+# AlgorithmMD
+ALG_MDGPS = {
+    'inner_iterations': 4,
```
Looks like this is unused. Remove it?

cbfinn commented on python/gps/algorithm/config.py:
```diff
@@ -44,3 +41,12 @@
+# AlgorithmMD
+ALG_MDGPS = {
+    'inner_iterations': 4,
+    'init_pol_wt': 0.01,
```
Is this also unused?

wmontgomery4 replied:
Yes, these are both unused, thanks for the catch!

wmontgomery4 replied:
Actually, I still need 'init_pol_wt', because the pol_info constructor requires it (and the policy_opt code uses pol_wt in several places). It doesn't matter for mdgps since the pol_wt remains constant, but it would be a good thing to separate out when refactor.

cbfinn replied:
Ok, sounds good. Maybe just add a note that it's algorithmically unused.

cbfinn commented on python/gps/gps_main.py:
```diff
@@ -19,7 +19,7 @@
 from gps.utility.data_logger import DataLogger
 from gps.sample.sample_list import SampleList
 
-logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)
```
might be nice to keep the default here as INFO.

wmontgomery4 replied:
Ok no problem. I've found DEBUG pretty useful, but I can just keep that as a local change.

cbfinn commented on python/gps/algorithm/traj_opt/traj_opt_lqr_python.py:
```diff
@@ -30,7 +32,19 @@ def update(self, m, algorithm):
+        if type(algorithm) == AlgorithmMDGPS:
+            # For MDGPS, constrain to previous NN linearization
+            # NOTE: we only copy in necessary params, ignoring inv_pol_covar
+            pol_info = algorithm.cur[m].pol_info
```
might be nice to put this in a helper function. It's rather verbose.

wmontgomery4 replied:
I was debating about that. My plan was to just add a `pol_info_to_traj_distr` function in `algorithm_utils.py`, but I decided against it because the proper thing would be to compute 'inv_pol_covar' and that's never needed. Definitely premature optimization though.. so I'll just add that helper method. In the refactor, I think we might want to add a "linearization" object which can come from a policy or LG controller. We could use that everywhere traj_distr or a neural net linearization is needed.

cbfinn replied:
Ok, sounds good!

cbfinn commented on python/gps/algorithm/traj_opt/traj_opt_lqr_python.py:
```diff
@@ -221,13 +239,20 @@ def backward(self, prev_traj_distr, traj_info, eta, algorithm, m):
-                    Qt = Qt + \
-                            Fm[t, :, :].T.dot(Vx[t+1, :] +
-                                              Vxx[t+1, :, :].dot(fv[t, :]))
+                    if type(algorithm) == AlgorithmBADMM:
```
Can we just write this as something like:

```
if type(algorithm) == AlgortihmBADMM:
    multiplier = (pol_wt[t+1] + eta)/(pol_wt[t] + eta)
else:
    multiplier = 1
Qtt = Qtt + multiplier * ...
```

I think this would be easier to read, and slightly less code. If you can think of a better variable name for multiplier, feel free to change.

wmontgomery4 replied:
Sure, sounds good.

cbfinn commented:
Generally looks good! Left some fairly minor comments. 

I'm think it would be okay to merge before refactoring with BADMM, and leave the refactor for the next PR, since this one is already rather large.

wmontgomery4 committed:
set up tf hyperparams

wmontgomery4 committed:
fixes from chelseas comments

wmontgomery4 committed:
cant actually remove init_pol_wt from mdgps yet

wmontgomery4 committed:
make policy update normalization automatic

wmontgomery4 committed:
helper method for getting traj_distr from pol_info

wmontgomery4 committed:
fixed tf bug, actually works with tf now

wmontgomery4 committed:
forgot to refit policy linearization after policy update

wmontgomery4 committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	python/gps/gui/gps_training_gui.py

wmontgomery4 committed:
add kl_div gui fixes to mdgps

wmontgomery4 committed:
set random seed explicitly in policy_opt_tf

wmontgomery4 committed:
clean up peg mdgps hyperparams file

wmontgomery4 committed:
fixed bizarre tensorflow import bug

wmontgomery4 commented:
Okay, I've addressed the comments and switched the hyperparams file over to using tensorflow. I need to tweak the hyperparams a bit, but everything should be essentially working.

One note, I changed the kl_div_i/f outputs to show the average kl_div rather than the sum, since this is what the `step` column corresponds to already. If you'd prefer the sum let me know and I'll change it back.

Also, I had a bizarre bug where tensorflow would seg fault unless I continued to import `policy_opt_caffe.py` (even though it was unused). I fixed it by changing the order of imports in `policy_opt_tf.py`, but I'm still not sure what the issue was (it might just have been a local issue).

cbfinn commented:
Ok, mean kl is fine. That's a rather weird tf bug.

I'll add you as a collaborator so you can merge whenever you're ready.

wmontgomery4 committed:
fix policy_opt bug for badmm

wmontgomery4 committed:
revert gmm convergence check
